Add SwPush notification usage to api doc

jbogarthyde opened the pull request:
## PR Checklist
Please check if your PR fulfills the following requirements:

- [ ] The commit message follows our guidelines: https://github.com/angular/angular/blob/master/CONTRIBUTING.md#commit
- [ ] Tests for the changes have been added (for bug fixes / features)
- [x] Docs have been added / updated (for bug fixes / features)


## PR Type
What kind of change does this PR introduce?

<!-- Please check the one that applies to this PR using "x". -->

- [ ] Bugfix
- [ ] Feature
- [ ] Code style update (formatting, local variables)
- [ ] Refactoring (no functional changes, no api changes)
- [ ] Build related changes
- [ ] CI related changes
- [x] Documentation content changes
- [ ] angular.io application / infrastructure changes
- [ ] Other... Please describe:


## What is the current behavior?
Developers have complained that doc is insufficient for using SwPush API.

Issue Number:  https://github.com/angular/angular/issues/22627


## What is the new behavior?
Adds usage information and links to further info to API doc:

https://angular.io/api/service-worker/SwPush

## Does this PR introduce a breaking change?

- [ ] Yes
- [x] No

jbogarthyde committed:
docs: add push notification usage to api doc

mary-poppins commented:
You can preview 632d937 at https://pr32139-632d937.ngbuilds.io/.

mary-poppins commented:
You can preview 84f882e at https://pr32139-84f882e.ngbuilds.io/.

AndrewKushnir commented:
@jbogarthyde this PR still requires an approval (i.e. can not be merged), right?

gkalpak commented on packages/service-worker/src/push.ts:
```diff
@@ -14,7 +14,79 @@ import {ERR_SW_NOT_SUPPORTED, NgswCommChannel, PushEvent} from './low_level';
+ *
+ * While the subscription is active, Service Worker listens for PushEvents and creates
+ * Notification instances in response.
+ *
```
You could, also, mention that the app can subscribe to `SwPush.notificationClicks` to get notified when a user clicks on a notification. E.g.:

```ts
swPush.notificationClicks.subscribe(({action, notification}) => {  
  // TODO: Do something in response to notification click.
});
```

gkalpak commented on packages/service-worker/src/push.ts:
```diff
@@ -14,7 +14,79 @@ import {ERR_SW_NOT_SUPPORTED, NgswCommChannel, PushEvent} from './low_level';
+ *     } catch (e) {
+ *       console.error('Could not subscribe', e);
+ *     }
+ *   }
```
Incorrect indentation (but hard-coded code snippets are evil anyway 😛).

gkalpak reviewed (changes requested):
My main issue is with the hard-coded code snippets.
Otherwise lgtm (with some minor comments).

mary-poppins commented:
You can preview 72af6ce at https://pr32139-72af6ce.ngbuilds.io/.

mary-poppins commented:
You can preview d7b1714 at https://pr32139-d7b1714.ngbuilds.io/.

mary-poppins commented:
You can preview 7ecd947 at https://pr32139-7ecd947.ngbuilds.io/.

mary-poppins commented:
You can preview 60943b6 at https://pr32139-60943b6.ngbuilds.io/.

mary-poppins commented:
You can preview 788eb8f at https://pr32139-788eb8f.ngbuilds.io/.

mary-poppins commented:
You can preview 6fdf984 at https://pr32139-6fdf984.ngbuilds.io/.

mary-poppins commented:
You can preview 1db81b0 at https://pr32139-1db81b0.ngbuilds.io/.

gkalpak commented on packages/service-worker/src/push.ts:
```diff
@@ -14,30 +14,26 @@ import {ERR_SW_NOT_SUPPORTED, NgswCommChannel, PushEvent} from './low_level';
- *
- *
- *
+ * Subscribe and listen to Web Push Notifications through Angular Service Worker.
```
Might be worth making `Web Push Notifications` a link to an explainer of push notifications (maybe [MDN: Web Push API Notifications best practices][1]).

[1]: https://developer.mozilla.org/en-US/docs/Web/API/Push_API/Best_Practices

gkalpak commented on packages/service-worker/src/push.ts:
```diff
@@ -14,7 +14,86 @@ import {ERR_SW_NOT_SUPPORTED, NgswCommChannel, PushEvent} from './low_level';
+ *     } catch (e) {
+ *       console.error('Could not subscribe', e);
+ *     }
+ *   }
```
Incorrect indentation (but hard-coded code snippets are evil anyway 😛).

gkalpak replied:
Setting up examples is a little involved (but totally worth it in the long run imo).
I'll try to put something together.

gkalpak commented on packages/service-worker/src/push.ts:
```diff
@@ -14,7 +14,86 @@ import {ERR_SW_NOT_SUPPORTED, NgswCommChannel, PushEvent} from './low_level';
+ *      "timestamp": DOMTimeStamp,
+ *      "title": DOMString,
+ *      "vibrate": number[]
+ *    }
```
We typically use an indentation of 2 spaces (not 3) 🤔

gkalpak replied:
Not quite :smiley: The first level has an indentation of 3 spaces instead of 2. (The second level is indentated correctly relative to its parent).

```diff
 {
-   "notification": {
-     "actions": NotificationAction[],
-     ...
+  "notification": {
+    "actions": NotificationAction[],
+    ...
   }
 }
```

I'll fix it up for you.

gkalpak commented on packages/service-worker/src/push.ts:
```diff
@@ -14,7 +14,86 @@ import {ERR_SW_NOT_SUPPORTED, NgswCommChannel, PushEvent} from './low_level';
+ *
+ * Unsubscribe using `SwPush.unsubscribe()`.
+ *
+ * An application can subscribe to `SWPush.notificationClicks` observable
```
`SWPush` --> `SwPush`

gkalpak commented on packages/service-worker/src/push.ts:
```diff
@@ -14,7 +14,86 @@ import {ERR_SW_NOT_SUPPORTED, NgswCommChannel, PushEvent} from './low_level';
+ * [instance properties](https://developer.mozilla.org/en-US/docs/Web/API/Notification#Instance_properties).
+ *
+ * While the subscription is active, Service Worker listens for a `PushEvent` occurences
+ * and creates `Notification` instances in response.
```
Worth linking to [PushEvent][1] and [Notification][2].

[1]: https://developer.mozilla.org/en-US/docs/Web/API/PushEvent
[2]: https://developer.mozilla.org/en-US/docs/Web/API/Notification

gkalpak committed:
fixup! docs: add push notification usage to api doc

gkalpak committed:
docs(service-worker): add example app for `SwPush` API docs

Previously, the `SwPush` API docs were using hard-coded code snippets.

This commit switches to using code snippets from an actual example app,
which ensures that the code shown in the docs will at least continue to
compile successfully.

gkalpak commented:
@googlebot I consent.

gkalpak commented:
**merge-assistance:** Global approval for this (more or less) docs-only change.
